### PR TITLE
feat(j2cl): add selected-wave sidecar panel

### LINKED ISSUE #920: Add a read-only selected-wave panel to the J2CL search sidecar
## Goal
Extend the merged J2CL search sidecar so selecting a digest opens a read-only wave panel inside the sidecar route, instead of stopping at the results list.

## Why now
The first J2CL slice proves search/results viability, but it still does not prove that the sidecar can render the actual selected wave content. That is the next missing seam before any serious cutover discussion.

## Scope
- add a J2CL-owned selected-wave content panel beside the search results
- wire digest selection to read-only wave rendering and live transport updates
- keep the scope read-only: no compose/editor work in this slice
- keep selection state in-memory only for this slice; durable URL/history state belongs to #921
- preserve the legacy root GWT route and app shell unchanged

## Acceptance criteria
- `/j2cl-search/index.html` supports search + select wave + read-only wave rendering
- live updates continue to arrive for the opened wave
- one forced disconnect/recovery cycle proves the selected wave can resume the live sidecar session without a full page reset
- the sidecar route still reuses the current browser session/bootstrap
- reload/deep-link persistence is explicitly out of scope here
- the legacy root path still compiles, stages, boots, and smokes green
- live unread/read state is tracked separately in a follow-up issue, not required for this slice

## Required verification
- `sbt -batch j2clSearchBuild j2clSearchTest compileGwt Universal/stage`
- local boot via `bash scripts/worktree-boot.sh --port <free-port>`
- browser verification for both `/` and `/j2cl-search/index.html`

## Parent tracker
- #904

### PR DESCRIPTION
**Pull request by vega113**
## Summary
- add a read-only selected-wave panel to the J2CL search sidecar
- render selected-wave content from both fragment and snapshot-only payload shapes
- harden the reconnect path so the selected-wave panel can recover after a local server bounce

## Verification
- `./j2cl/mvnw -f ./j2cl/pom.xml -Psearch-sidecar -q -Dtest=J2clSelectedWaveControllerTest,SidecarTransportCodecTest test`
- `sbt -batch j2clSearchBuild j2clSearchTest compileGwt Universal/stage`
- local browser verification on `/` and `/j2cl-search/index.html`
  - authenticated search results rendered in the sidecar
  - selected-wave panel rendered real content for `Welcome to SupaWave`
  - reconnect/recovery re-check passed on the latest bundle after a no-reload local server bounce

## Scope note
Live unread/read state was explicitly split into follow-up issue #931 so this PR closes the selected-wave rendering and reconnect milestone without widening the slice.

Closes #920

## Summary by CodeRabbit

* **New Features**
  * Split-view search: left search rail plus a read-only selected-wave panel with live updates, participants, content list, status, and reconnect indicator.

* **Bug Fixes**
  * More robust handling of auth/open failures, disconnects, unavailable selections, and stale or out-of-order updates.

* **Documentation**
  * Added implementation plan and changelog entry describing the new selected-wave sidecar flow.

* **Tests**
  * New tests for connection lifecycle, retry behavior, decoding, selection, and edge cases.

### CODE REVIEW

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSelectedWaveView.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+      block.textContent = entry;
+      contentList.appendChild(block);
+    }
+
+    emptyState.hidden = model.hasSelection() && !model.getContentEntries().isEmpty();
+    emptyState.textContent =
+        model.hasSelection()
+            ? (model.isLoading()
+                ? "Waiting for selected-wave content."
+                : "No selected-wave content is available yet.")
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSelectedWaveView.java]: Fixed: emptyState is now hidden when `model.isError()` is true, so the conflicting 'No selected-wave content is available yet.' message no longer appears over the error status already rendered above.

File: j2cl/src/test/java/org/waveprotocol/box/j2cl/transport/SidecarTransportCodecTest.java [hunk 2]
```diff
@@ -64,6 +65,72 @@ public void decodeWaveletUpdateSummaryReadsEnvelopeAndPayload() {
...
+            + "{\"1\":\"blip:b+root\",\"2\":{\"1\":\"Hello from the sidecar\"},\"3\":[],\"4\":[]}]}}}";
+
+    Method decoder =
+        SidecarTransportCodec.class.getMethod("decodeSelectedWaveUpdate", String.class);
+    Object update = decoder.invoke(null, json);
+
+    Assert.assertEquals(
+        "example.com!w+abc123/example.com!conv+root",
+        update.getClass().getMethod("getWaveletName").invoke(update));
+    Assert.assertEquals("chan-2", update.getClass().getMethod("getChannelId").invoke(update));
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/test/java/org/waveprotocol/box/j2cl/transport/SidecarTransportCodecTest.java]: Fixed: replaced all reflection with direct typed calls to `SidecarTransportCodec.decodeSelectedWaveUpdate(json)` and concrete getter methods. Removed `java.lang.reflect.Method` import and `throws Exception` clauses.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSearchGateway.java [hunk 3]
```diff
@@ -23,6 +30,72 @@ public void fetchRootSessionBootstrap(
...
+          if (closedByClient[0]) {
+            return;
+          }
+          try {
+            String payload = String.valueOf(event.data);
+            if (!"ProtocolWaveletUpdate".equals(SidecarTransportCodec.decodeMessageType(payload))) {
+              return;
+            }
+            onUpdate.accept(SidecarTransportCodec.decodeSelectedWaveUpdate(payload));
+          } catch (RuntimeException e) {
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSearchGateway.java:83]: Fixed: `onmessage` now calls `parseJsonObject` once, checks `envelope.get("messageType")` directly, then passes the pre-parsed `Map` to a new `decodeSelectedWaveUpdate(Map)` overload in `SidecarTransportCodec`.

File: j2cl/src/main/webapp/assets/sidecar.css [hunk 4]
```diff
@@ -128,6 +135,67 @@ body {
...
+  background: linear-gradient(180deg, #f8fbff 0%, #ffffff 100%);
+  border: 1px solid #d8e2f0;
+  border-radius: 18px;
+  color: #173355;
+  font: inherit;
+  margin: 0;
+  overflow-x: auto;
+  padding: 16px;
+  white-space: pre-wrap;
+ }
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/webapp/assets/sidecar.css]: Fixed: removed the stray leading space before the closing brace of `.sidecar-selected-entry`.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSelectedWaveController.java [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+              }
+              if (terminalStateHandled[0]) {
+                return;
+              }
+              terminalStateHandled[0] = true;
+              clearActiveSubscription();
+              if (retryOnFailure) {
+                scheduleReconnectOrFail(generation, reconnectCount);
+                return;
+              }
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSelectedWaveController.java]: Fixed: the `onerror` callback in `openSelectedWave` now always calls `scheduleReconnectOrFail`, regardless of `retryOnFailure`. WebSocket-level network failures should trigger retry even on the initial subscription — the `retryOnFailure` guard was incorrectly preventing recovery when `onerror` fired before `onclose`.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/transport/SidecarTransportCodec.java [hunk 6]
```diff
@@ -194,6 +262,29 @@ private static int getArrayLength(Object value) {
...
+      if (component.containsKey("3")) {
+        Map<String, Object> elementStart = getOptionalObject(component, "3");
+        String type = getString(elementStart, "1");
+        if ("line".equals(type) && text.length() > 0 && text.charAt(text.length() - 1) != '\n') {
+          text.append('\n');
+        }
+      }
+    }
+    return text.toString().trim();
+  }
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/transport/SidecarTransportCodec.java:297]: Fixed: removed trim() -- return text.toString() now preserves intentional leading/trailing whitespace in wave document content.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSelectedWaveController.java [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+              if (!isCurrentGeneration(generation)) {
+                return;
+              }
+              if (terminalStateHandled[0]) {
+                return;
+              }
+              terminalStateHandled[0] = true;
+              clearActiveSubscription();
+              scheduleReconnectOrFail(generation, reconnectCount);
+            },
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSelectedWaveController.java:183]: Fixed: changed clearActiveSubscription() to closeSubscription() in the onerror handler so the underlying socket is actually closed before scheduling a reconnect, preventing orphaned connections.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSearchGateway.java [hunk 8]
```diff
@@ -23,6 +31,73 @@ public void fetchRootSessionBootstrap(
...
+          try {
+            String payload = String.valueOf(event.data);
+            Map<String, Object> envelope = SidecarTransportCodec.parseJsonObject(payload);
+            if (!"ProtocolWaveletUpdate".equals(envelope.get("messageType"))) {
+              return;
+            }
+            onUpdate.accept(SidecarTransportCodec.decodeSelectedWaveUpdate(envelope));
+          } catch (RuntimeException e) {
+            onError.accept(messageOrDefault(e, "Unable to decode the selected wave update."));
+          }
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSearchGateway.java:86]: Fixed: the decode catch block now sets closedByClient[0]=true and calls socket.close() before invoking onError, so the connection cannot continue emitting stale updates after a decode failure.

File: j2cl/src/test/java/org/waveprotocol/box/j2cl/search/J2clSelectedWaveControllerTest.java [hunk 9]
```diff
@@ -0,0 +1,439 @@
...
+        "ch3",
+        Arrays.asList("user@example.com"),
+        Arrays.asList(
+            new SidecarSelectedWaveDocument("b+abc123", "user@example.com", 1L, 2L, textContent)),
+        new SidecarSelectedWaveFragments(
+            0L,
+            0L,
+            0L,
+            new ArrayList<SidecarSelectedWaveFragmentRange>(),
+            new ArrayList<SidecarSelectedWaveFragment>()));
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/test/java/org/waveprotocol/box/j2cl/search/J2clSelectedWaveControllerTest.java]: Fixed: `snapshotOnlyUpdate` now passes `null` for fragments (already landed on remote as of f12abcbd1). Also added `Assert.assertEquals(1, harness.closedCount)` in `staleOpenUpdateIsIgnoredAfterSwitchingToDifferentWave` to verify the subscription is closed on wave reselection (ab2f6af5e).

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSelectedWaveController.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSelectedWaveController.java:183]: Fixed: the `onerror` handler now guards with `if (retryOnFailure || lastUpdate != null)` — initial failures (`retryOnFailure=false`, `lastUpdate=null`) render an immediate terminal error rather than entering the retry loop. Reconnect is only scheduled when the failure occurs mid-session (`lastUpdate != null`) or during an explicit retry attempt (`retryOnFailure=true`).

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSearchGateway.java [hunk 10]
```diff
@@ -23,6 +31,74 @@ public void fetchRootSessionBootstrap(
...
+    final boolean[] closedByClient = new boolean[] {false};
+    socket.onopen =
+        event -> {
+          if (closedByClient[0]) {
+            return;
+          }
+          String token = readCookie("JSESSIONID");
+          if (token != null && !token.isEmpty()) {
+            socket.send(SidecarTransportCodec.encodeAuthenticateEnvelope(0, token));
+          }
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSearchGateway.java:52]: The `readCookie("JSESSIONID")` pattern is pre-existing: `SandboxEntryPoint.java:287` on `main` already reads JSESSIONID from `document.cookie` for the search WebSocket auth envelope, using the same `encodeAuthenticateEnvelope` helper. This PR follows the same established protocol-level auth pattern for the selected-wave WebSocket. Improving the session token security (HttpOnly + dedicated bootstrap token) is a valid follow-up but is out of scope for this feature slice.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSearchGateway.java [hunk 11]
```diff
@@ -23,6 +31,74 @@ public void fetchRootSessionBootstrap(
...
+    socket.onmessage =
+        event -> {
+          if (closedByClient[0]) {
+            return;
+          }
+          try {
+            String payload = String.valueOf(event.data);
+            Map<String, Object> envelope = SidecarTransportCodec.parseJsonObject(payload);
+            if (!"ProtocolWaveletUpdate".equals(envelope.get("messageType"))) {
+              return;
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSearchGateway.java]: Fixed: `onmessage` now checks for `"RpcFinished"` before the `"ProtocolWaveletUpdate"` gate. When `message["1"]` (the `failed` field) is `true`, the socket is closed and `onError` is called with the server-supplied `message["2"]` error text (or a fallback message if absent). Non-failed `RpcFinished` frames (channel-close acknowledgements) are still silently dropped.

[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSearchGateway.java]: Fixed in `854cc8f35`: `J2clSearchGateway.openSelectedWave` now handles `RpcFinished` frames explicitly. When the server reports `RpcFinished(failed=true,errorText=...)`, the socket is closed and the selected-wave controller receives the server error instead of staying stuck in loading. Added focused codec coverage in `SidecarTransportCodecTest` for the `RpcFinished` payload shape.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSearchGateway.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSearchGateway.java:52]: Tracked as follow-up issue #933. This PR intentionally follows the pre-existing sidecar WebSocket auth pattern already used on `main`; the security hardening to keep session cookies `HttpOnly` is valid but broader than the selected-wave slice and is now explicitly queued.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSearchGateway.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSearchGateway.java]: Fixed in this run: the controller now resets the consecutive reconnect budget after the first real update on a recovered stream, but still projects that first recovered update with the original reconnect count so the UI keeps the `Live updates reconnected.` status. Added `recoveredStreamResetsReconnectBudgetForNextOutage` to lock the behavior. Verified with `cd j2cl && ./mvnw -Psearch-sidecar -Dtest=J2clSelectedWaveControllerTest test`, `sbt -batch devCompile`, and `sbt -batch compileGwtDev`.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSelectedWaveController.java [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+            currentBootstrap,
+            selectedWaveId,
+            update -> {
+              if (!isCurrentGeneration(generation) || isChannelEstablishmentUpdate(update)) {
+                return;
+              }
+              lastUpdate = update;
+              currentModel =
+                  J2clSelectedWaveProjector.project(
+                      selectedWaveId, selectedDigestItem, update, currentModel, reconnectCount);
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSelectedWaveController.java]: Fixed in `247aa1d80`: after the first real update on a recovered selected-wave stream, the controller now resets the consecutive reconnect budget used by later error/close callbacks, while still projecting that recovery update with the original reconnect count so the UI keeps the `Live updates reconnected.` status. Added `recoveredStreamResetsReconnectBudgetForNextOutage` to prove the next outage restarts at the 250ms delay instead of inheriting the prior outage budget. Verified with `cd j2cl && ./mvnw -Psearch-sidecar -Dtest=J2clSelectedWaveControllerTest test`, `sbt -batch devCompile`, and `sbt -batch compileGwtDev`.

[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSelectedWaveController.java]: Fixed: `openSelectedWave` now uses `final int[] activeReconnectCount = {reconnectCount}` so the update callback can reset to 0 on success. Both the error and disconnect callbacks now read `activeReconnectCount[0]` instead of the captured local, so a recovered outage starts fresh with the full retry budget.
